app/src/tiles/dataDefinition.ts

```diff
@@ -0,0 +1,135 @@
+            g.geometry_geom
+        FROM
+            geometries as g,
+            buildings as b
+        WHERE
+            g.geometry_id = b.geometry_id
+    ) as outline`,
+    size_storeys: `(
+        SELECT
+            (
+                coalesce(b.size_storeys_attic, 0) +
+                coalesce(b.size_storeys_core, 0)
+            ) as size_storeys,
+            g.geometry_geom
+        FROM
+            geometries as g,
+            buildings as b
+        WHERE
+            g.geometry_id = b.geometry_id
+    ) as outline`,
+    location: `(
+        SELECT
+            (
+                case when b.location_name is null then 0 else 1 end +
+                case when b.location_number is null then 0 else 1 end +
+                case when b.location_street is null then 0 else 1 end +
+                case when b.location_line_two is null then 0 else 1 end +
+                case when b.location_town is null then 0 else 1 end +
+                case when b.location_postcode is null then 0 else 1 end +
+                case when b.location_latitude is null then 0 else 1 end +
+                case when b.location_longitude is null then 0 else 1 end +
+                case when b.ref_toid is null then 0 else 1 end +
+                case when b.ref_osm_id is null then 0 else 1 end
+            ) as location_info_count,
+            g.geometry_geom
+        FROM
+            geometries as g,
+            buildings as b
+        WHERE
+            g.geometry_id = b.geometry_id
+    ) as location`,
+    likes: `(
+        SELECT
+            g.geometry_geom,
+            b.likes_total as likes
+        FROM
+            geometries as g,
+            buildings as b
+        WHERE
+            g.geometry_id = b.geometry_id
+            AND b.likes_total > 0
+    ) as location`,
+    conservation_area: `(
+        SELECT
+            g.geometry_geom
+        FROM
+            geometries as g,
+            buildings as b
+        WHERE
+            g.geometry_id = b.geometry_id
+            AND b.planning_in_conservation_area = true
+    ) as conservation_area`
+};
+
+const GEOMETRY_FIELD = 'geometry_geom';
+
+function getBuildingsDataConfig(tileset: string, dataParams: any): DataConfig {
+    const table = BUILDING_LAYER_DEFINITIONS[tileset];
+
+    if(table == undefined) {
+        throw new Error('Invalid tileset requested');
+    }
+
+    return {
+        geometry_field: GEOMETRY_FIELD,
+        table: table
+    };
+}
+
+function getHighlightDataConfig(tileset: string, dataParams: any): DataConfig {
+    let { highlight, base } = dataParams;
+
+    highlight = strictParseInt(highlight);
+    base = base || 'default';
+
+    if(isNaN(highlight) || base.match(/^\w+$/) == undefined) {
+        throw new Error('Bad parameters for highlight layer');
+    }
+
+    return {
+        geometry_field: GEOMETRY_FIELD,
+        table: `(
+            SELECT
+                g.geometry_geom,
+                '${base}' as base_layer
```

Comment:
I'm not sure that would be better - the goal is to keep the style-related configuration in one place, and that would be the XML. Instead of `highlight_red` we could create tilesets like `highlight_location` etc. But that requires more work whenever you want to change something, because you need to manage too many tilesets, and it makes it difficult to have configurations like default highlight style + custom style for just a few base layers.

The passing through SQL is not the cleanest, ideally mapnik would accept render-time variables (https://github.com/mapnik/mapnik/issues/2553) but otherwise I think this is the solution with the highest scalability and separation of concerns. I think we just need to be careful not to introduce any SQL injection vulnerabilities.

Reply:
Created issue #438 to follow up on the OO style of the renderer. Will merge this for now and will refactor post launch